Add touch support in our carousel

Johann-S opened the pull request:
- [x] Add unit test
- [x] Documentation

Fixes: #17118

Johann-S committed:
Add touch support in our carousel with HammerJS.

XhmikosR commented:
Its quite a big addition... Can't we do the basic stuff ourselves?

XhmikosR commented on _includes/scripts.html:
```diff
@@ -2,6 +2,8 @@
 <script>window.jQuery || document.write('<script src="{{ site.baseurl }}/assets/js/vendor/jquery-slim.min.js"><\/script>')</script>
 
 <script src="{{ site.baseurl }}/assets/js/vendor/popper.min.js"{% if site.github %} integrity="{{ site.cdn.popper_hash }}" crossorigin="anonymous"{% endif %}></script>
+<script src="{{ site.baseurl }}/assets/js/vendor/hammer.min.js"{% if site.github %} integrity="{{ site.cdn.hammer_hash }}" crossorigin="anonymous"{% endif %}></script>
```
You need to add it in package.json too for the docs package.

Johann-S replied:
It's strange but Popper isn't in our `doc-minify` task 🤔

XhmikosR replied:
True, indeed. Just leave it as is and I'll give it a go in the next days and let you know. :)

Johann-S commented:
I don't think it's a big change here, we add an optionnal lib, if folks need a touch support that's all

XhmikosR commented:
Sure, but can't we have the basic functionality without a 20 KB file? I recall you had a branch in the past.

Johann-S commented:
This won't be added in our dist files, so we won't add 20 kb

I'm not in favor of adding logics about touch detection and swipe detection when they are libs which can do that better than me

XhmikosR commented:
I agree on that, but then one would say why don't you support jQuery mobile or whatever other lib there's out there.

Anyway, I'm not against this, I just want us to think it through.

Johann-S commented:
I think HammerJS is a good choice, Angular chose HammerJS too

jackmu95 commented on js/src/carousel.js:
```diff
@@ -244,16 +261,22 @@ const Carousel = (($) => {
       }
 
+      if (touchSupported) {
+        this.hammer.on(Event.SWIPELEFT, () => this.next())
```
Isn't here a check quired if hammer support is enabled or the hammer reference exists to prevent a reference error? 🤔

Johann-S replied:
Yep that's true nice catch 👍

Johann-S commented on js/tests/index.html:
```diff
@@ -20,11 +20,15 @@
       }())
     </script>
     <script src="../../assets/js/vendor/popper.min.js"></script>
+    <script src="../../assets/js/vendor/hammer.min.js"></script>
```
yep sure !

XhmikosR commented on js/tests/karma.conf.js:
```diff
@@ -17,6 +17,8 @@ module.exports = (config) => {
     files: [
       jqueryFile,
       'assets/js/vendor/popper.min.js',
+      'assets/js/vendor/hammer.min.js',
```
This shouldn't be needed.

XhmikosR commented:
@Johann-S can you rebase this?

Johann-S commented:
Done @XhmikosR 👍

Johann-S commented:
I cannot validate our test because on SauceLabs we run our test on Linux Firefox 45, Linux Chrome 48 and Mac OS X 10.11 Safari 10 which are obsolete version

XhmikosR commented:
@Johann-S: so we can't merge this now if we can't test it, right? What can we do about this?

XhmikosR commented:
@mdo: can this be done in 4.2?

Johann-S commented:
@XhmikosR everything is done here and works, thanks to BrowserStack 😄 

/CC @bardiharborow for a review

patrickhlauke commented on js/tests/unit/carousel.js:
```diff
@@ -1,6 +1,8 @@
   'use strict'
 
+  var touchSupported = 'ontouchstart' in document.documentElement
+
```
for testing purposes, this is of course fine. i'll just note here that to detect pointer events touch support you'd want to also check for `navigator.maxTouchPoints > 0` (in production on a live site, not for automated unit tests)

Johann-S replied:
yep, I added that line because on Safari our unit tests with touch events failed 😄

patrickhlauke commented on site/docs/4.1/components/carousel.md:
```diff
@@ -12,6 +12,8 @@ The carousel is a slideshow for cycling through a series of content, built with
 
 In browsers where the [Page Visibility API](https://www.w3.org/TR/page-visibility/) is supported, the carousel will avoid sliding when the webpage is not visible to the user (such as when the browser tab is inactive, the browser window is minimized, etc.).
 
+The carousel can handle swipe events (left and right) only if you include [HammerJS]({{ site.cdn.hammer }} before Bootstrap.
```
perhaps reword this a tad:

"The carousel supports swipe gestures (left and right) using <a href="https://hammerjs.github.io/">HammerJS</a>. For this to function correctly you need to include [HammerJS]...."

Johann-S replied:
I changed that, thanks Patrick !

XhmikosR commented:
@Johann-S: can you rebase this one please? :)

Johann-S commented:
Done @XhmikosR 😉

XhmikosR commented:
All good with this @Johann-S. Only thing I was thinking is if we should have an option to enable/disable touch support.

Johann-S commented:
If they don't include HammerJS there is no touch support

XhmikosR commented:
Yes, but the bundle already includes it.

Johann-S commented on js/src/carousel.js:
```diff
@@ -96,11 +101,23 @@ const Carousel = (($) => {
       this._element            = $(element)[0]
       this._indicatorsElement  = this._element.querySelector(Selector.INDICATORS)
 
+      if (HAMMER_ENABLED) {
```
I should use hammer only if touch events supported

ben29 commented:
So when it will be supported?

XhmikosR commented:
@Johann-S: would you be so kind and rebase this one last time (hopefully)?

Johann-S commented:
Done @XhmikosR 😉 Maybe someday it'll be merged 🙏

XhmikosR commented:
OK, I think this is ready to be merged :)

XhmikosR commented on site/docs/4.1/components/carousel.md:
```diff
@@ -281,6 +283,12 @@ Options can be passed via data attributes or JavaScript. For data attributes, ap
+    <tr>
+      <td>touch</td>
+      <td>boolean</td>
+      <td>true</td>
```
One thing. This is true, but maybe we should mention that it will be disabled gracefully if HammerJs isn't present?

Johann-S replied:
You're right we should add that

Johann-S committed:
swipe left/right without hammerjs

patrickhlauke reviewed (dismissed):
the documentation part may need some more clarity, as it's not quite obvious to me what that even means...

patrickhlauke commented on site/docs/4.1/components/carousel.md:
```diff
@@ -287,7 +285,7 @@ Options can be passed via data attributes or JavaScript. For data attributes, ap
       <td>boolean</td>
       <td>true</td>
-      <td>Whether the carousel should handle touch event and allow swipe left/right.</td>
+      <td>Whether the carousel should handle touch events and allow swipe left/right (only for mobile browsers).</td>
```
no, not really "only mobile browsers" - there are desktop computers with touch (e.g. surface). also, hammer uses pointer events where available, so technically not "touch events". suggest changing it to what i said originally in my follow-up comment :)

Johann-S replied:
Oh ok I'll change that. We removed HammerJS we use a Vanilla JS solution

patrickhlauke replied:
oh, missed that you're using vanilla JS (as i saw the hammer simulator being loaded as well). in which case could I'd say it would be good to also support pointer events...but that's probably a more involved addition...

patrickhlauke replied:
for reference: supporting pointer events will also make this work on touch-enabled desktop/laptop devices in Edge (which, by default, does not expose touch events by default, but does support touch interactions via pointer events)

patrickhlauke commented:
do any of the demo carousels on https://deploy-preview-25776--twbs-bootstrap4.netlify.com/docs/4.1/components/carousel/ have touch enabled? if not, would be good to do it and note it

Johann-S commented:
All the carousels have touch enabled 👍

patrickhlauke commented on js/src/carousel.js:
```diff
@@ -84,21 +89,21 @@ const Selector = {
+    this._config            = this._getConfig(config)
+    this._element           = element
+    this._indicatorsElement = this._element.querySelector(Selector.INDICATORS)
+    this._touchSupported    = 'ontouchstart' in document.documentElement || navigator.maxTouchPoints > 0
```
if you're not using pointer events, why are you testing for `navigator.maxTouchPoints` (which is part of the pointer events spec) ?

patrickhlauke commented:
swiping left/right in Chrome on desktop/laptop with touchscreen and on android also triggers back/forward navigation. it likely needs a preventDefault() in there somewhere (this is the sort of stuff that i think Hammer would have handled by default already, as well as pointer events support)

XhmikosR commented:
>the swipe direction seems counterintuitive. if i swipe right, i would want the carousel to go to the previous slide, not the next slide, and vice-versa

Yeah, that was one of the things that didn't feel right to me either. I asked Johann to keep the HammerJS patch just in case we need to revert. So if things tend to be too broken, we can still just go with Hammer.

patrickhlauke reviewed (changes requested):
- revert the swipe direction (so swiping left goes to next slide, not previous, and vice-versa)
- prevent swipes from triggering browser back/forward navigation (e.g. Chrome on Android / on Surface)

Johann-S commented:
it should be better @patrickhlauke can you confirm it ? Please

patrickhlauke commented:
strangely, it still doesn't work in Edge on Surface (while the hammer.js demo itself http://hammerjs.github.io/ does). i'll have a look later tonight to see if i can work out why. other than than, the swipe direction now feels natural, and back/forward in Chrome is suppressed when swiping. :+1:

Johann-S commented:
I tested (thanks to BrowserStack) on a Lumia 930 with Windows Phone 8.1 and it works fine 🤔

patrickhlauke commented:
Win Phone 8 behaves differently from a desktop Windows 10 on a touchscreen.

patrickhlauke commented:
ah, i see you had added hammer, but then removed it again, hence my confusion. to be clear: Edge/Win 10 on a touch-enabled desktop/laptop does not have touch events enabled, but does allow touch interaction via pointer events (filtered to `type = "touch"`, and with addition of relevant `touch-action: pan-x` to the element/container with the event handlers). as you reverted to only touch events again after removing hammer, it's not surprising it's not working in Edge on Surface. i'd say although it's an edge case (hah) at this point, since pointer events are the more modern way of handling things (and finally even Apple have them "in development" for future Safari), it's worth future-proofing this code here as well. i can give it a shot if you don't mind me fuzzing around with your branch?

patrickhlauke commented:
should ideally use both, with preference for PE where available. i'll have a go either later tonight or tomorrow

Johann-S committed:
use pointer events if available

Johann-S commented:
I implemented a solution with PointerEvent (finally I found the time 😄) but I'm not sure if I'm fully compatible 😟 
If you can help me on that @patrickhlauke or improve my code do not hesitate 👍

patrickhlauke commented on js/src/carousel.js:
```diff
@@ -230,23 +261,74 @@ class Carousel {
-      if ('ontouchstart' in document.documentElement) {
+    }
+
+    this._addTouchEventListeners()
```
should this not only be added if the configuration for touch is true?

patrickhlauke commented on js/src/carousel.js:
```diff
@@ -230,23 +261,74 @@ class Carousel {
+    this._addTouchEventListeners()
+  }
+
+  _addTouchEventListeners() {
```
beyond adding the event listener, this code should also add `touch-action: pan-x` to the carousel when pointer events are used ... `preventDefault()` doesn't stop the forward/back browser navigation for pointer events, it's done declaratively with `touch-action` in the PE model. either doing it dynamically (adding a `style`? seems a bit dirty) or adding it to the definition of carousel CSS styles (though then it would "swallow" horizontal swipes in a PE browser even if the author DIDN'T want touch to be handled on a carousel)

Johann-S replied:
I added a class when we use PointerEvents but I don't understand why it doesn't work on IE 10/11, do you any ideas why @patrickhlauke ? 🤔

patrickhlauke replied:
i'm an idiot...and suggested the wrong `touch-action` value...i should go and have a word with the co-editor and chair of the PE spec, it seems...

patrickhlauke commented on js/src/carousel.js:
```diff
@@ -230,23 +261,74 @@ class Carousel {
+      event.preventDefault()
+
+      if (this._pointerEvent) {
+        this.touchStartX = event.originalEvent.clientX
```
ideally also check first that the pointer event is of `pointerType=="touch"` - and possibly even `pointerType="pen"` if we want to allow pen/stylus swiping - as otherwise the swipe will also happen for mouse users. see https://w3c.github.io/pointerevents/#dom-pointerevent-pointertype

patrickhlauke commented on js/tests/unit/carousel.js:
```diff
@@ -989,4 +1018,189 @@ $(function () {
     }, 80)
   })
+
+  QUnit.test('should allow swiperight and call prev with pointer events', function (assert) {
```
does this only simulate/test touch events? or does it also test if the code works correctly for pointer events?

patrickhlauke committed:
Use correct touch-action values

- my fault, my original advice of using `touch-action: pan-x` is exactly the value we *don't* want to have the browser handle...

patrickhlauke commented:
urgh, i'm an idiot . the touch action that's needed isn't `touch-action: pan-x` ... i seem to always forget that `touch-action` really means "which actions should the *browser* handle", and horizontal panning is exactly what we *don't* want the browser to handle...

long story short, changing to `touch-action: pan-y pinch-zoom` makes this work in Edge/Surface and IE11/Surface as well

patrickhlauke committed:
Abandon swipe altogether if more than one touch detected

XhmikosR commented:
So in Firefox 62 dev tools emulation now it throws this:

```
TypeError: originEvent.touches is undefined carousel.js:287:8
start
carousel.js:287:8
_addTouchEventListeners/<
carousel.js:329:56
```

patrickhlauke committed:
Remove unnecessary pointer event listeners

these may also be the cause of weird behavior in Chrome/Surface, where scrolling vertically triggers slide advance

patrickhlauke committed:
Refactor (and correct) start/move/end functions

in particular, no need to use originEvent, and preventDefault() only needed for touch events

patrickhlauke committed:
Set touch-action to "none"

Firefox currently seems extremely fickle - with `pan-y` if fires pointercancel as soon as a touch strays even a pixel or so vertically.
While `touch-action: pan-y` would be ideal (allowing users to scroll the page even when their finger started the scroll on the carousel), this prevents a swipe that isn't perfectly/only horizontal to be recognised by Firefox.

Johann-S commented:
It seems not @patrickhlauke our unit tests do not pass on IE 10/11 😟

Johann-S commented:
@patrickhlauke I cleaned a bit the git history and squashed some of your commits

Johann-S committed:
avoid drag img

Johann-S commented:
@patrickhlauke everything is fine now, can you confirm it ? Thanks 👍

patrickhlauke commented on js/src/carousel.js:
```diff
@@ -114,8 +116,8 @@ class Carousel {
     this._indicatorsElement = this._element.querySelector(Selector.INDICATORS)
-    this._touchSupported    = 'ontouchstart' in document.documentElement || navigator.maxTouchPoints > 0
     this._pointerEvent      = Boolean(window.PointerEvent || window.MSPointerEvent)
+    this._touchSupported    = 'ontouchstart' in document.documentElement || this._pointerEvent
```
the original check for `navigator.maxTouchPoints > 0` was, i'd say, more appropriate. as pointer events (in modern browsers) are supported/present even if there's just a mouse, so you end up with unnecessary code being executed...unless you're doing this on purpose to cover the scenario where an (external) touch-capable display is attached at a later point after the page has already loaded, which is a bit niche...

patrickhlauke replied:
i guess if you did revert back to the `navigator.maxTouchPoints` check, you'd then need to also fake it in the unit tests so that that code gets run

Johann-S replied:
The unit tests wasn't in success because when you swipe on a Imagine with IE it start a drag event 😬 I'll to check maxTouchPoints and see if everything is fine

Johann-S replied:
Ok I restored the use of `navigator.maxTouchPoints`, it should be fine now 👍 I'll make some tests on mobile devices

Johann-S commented:
It works fine for me 👍 

/CC @XhmikosR if you want to do some tests too

patrickhlauke commented:
https://deploy-preview-25776--twbs-bootstrap4.netlify.com/docs/4.1/components/carousel/ seems to be working fine now (tested there as i've not got my build/test environment all set up again at the moment) ... but man, the aggressive caching/service worker really gets in the way of quickly checking changes out sometimes...

possibly more related to netlify here: on IE11 and Edge, I found that the carousels are unresponsive for quite a long time after page load (unless you kick them into action if there are the "<" / ">" controls). it seems the browser is waiting for some file to load from the network, and only once it gives up does it run the onload stuff.

XhmikosR commented:
There is actually no service worker on netlify; it doesn't run the production scripts on purpose.

patrickhlauke commented:
> There is actually no service worker on netlify; it doesn't run the production scripts on purpose.

i found i had to force refresh (and in the case of Firefox, go in and delete session storage etc) while testing. not quite sure why then...aggressive caching? i see pwa.js is loaded as well if that means anything...i'm actually not sure if this is related to service worker or not.

in any case, it seems to be the carbon ads json that is never loaded (shows as "(pending)" in the F12 tools' network tab). only once the browser gives up trying to load it is the JS on page load actually executed.

anyway, that's all tangential. the carousel swipe thing now seems to work nicely

XhmikosR committed:
Merge branch 'v4-dev' into v4-carousel-hammer

XhmikosR commented:
This no longer works for me with Firefox 63 devtools and say S9 emulation. It works fine with Chrome though.

Johann-S commented:
@XhmikosR that's strange it works perfectly on my Android with Firefox

XhmikosR commented:
Well a real touch device is not the same as the emulator :) That being said I don't know if we should even care about it. I just thought I'd report it since Firefox is my main browser and this worked in the previous patches.

Johann-S commented:
I think we should test using BrowserStack here

XhmikosR committed:
Merge branch 'v4-dev' into v4-carousel-hammer

XhmikosR commented:
All right, let's merge this then. Great work everyone!

XhmikosR committed:
Merge branch 'v4-dev' into v4-carousel-hammer
